File: src/modules/discord.py

import json, requests

def _call_method (method, bot_token, data = None, files = None):
	return requests.post(
		'https://discord.com/api/v10/{}'.format(method),
		headers = {
			'Authorization': 'Bot {}'.format(bot_token),
		},
		data = data,
		files = files,
	)

def send (notifier = None, text = None, log_file_name = None):
	chat_id = notifier['chat_id']
	bot_token = notifier['bot_token']

	api_method = 'channels/{}/messages'.format(chat_id)

	response = _call_method(
		api_method,
		bot_token = bot_token,
		data = {
			'content': text,
		},
	)
	# print('[send message] status code:', response.status_code)
	# print('[send message] body:', response.json())

	message_id = response.json()['id']

	if log_file_name:
		files = {
			'0': (
				'log.txt',
				open(log_file_name, 'rb'),
			),
		}

		response = _call_method(
			api_method,
			bot_token = bot_token,
			data = {
				'payload_json': json.dumps({
					'message_reference': {
						'message_id': message_id,
					},
				}),
			},
			files = files,
		)
		# print('[send file] status code:', response.status_code)
		# print('[send file] body:', response.json())

File: src/run/notification/discord.js

import { file_name as log_file_name } from '../../run/logger.file.js';
import { httpPostRequest }            from '../../utils.js';

async function callAPI(
	method,
	bot_token,
	data = {},
	files = {},
) {
	return httpPostRequest({
		url: `https://discord.com/api/v10/${method}`,
		headers: {
			Authorization: `Bot ${bot_token}`,
		},
		data,
		files,
	});
}

export default async function (
	{
		bot_token,
		chat_id,
	},
	text,
	add_log = false,
) {
	const api_method = `channels/${chat_id}/messages`;

	const response_message = await callAPI(
		api_method,
		bot_token,
		{
			content: text,
		},
	);
	if (process.env.SUPERNOVA_DEBUG) {
		console.log('response_message', response_message);
	}

	if (add_log) {
		const message_id = response_message.body.id;

		const response_file = await callAPI(
			api_method,
			bot_token,
			{
				payload_json: JSON.stringify({
					message_reference: {
						message_id,
					},
				}),
			},
			{
				0: {
					path: log_file_name,
					name: 'log.txt',
				},
			},
		);
		if (process.env.SUPERNOVA_DEBUG) {
			console.log('response_file', response_file);
		}
	}
}
